src/scripts/external/mastercrafted-api.ts

```diff
@@ -0,0 +1,235 @@
+    id?: string;
+    name?: string;
+    description?: string;
+    time?: number;
+    macroName?: string;
+    recipeBook?: RecipeBookMasterCrafted;
+    ingredients?: IngredientMasterCrafted[];
+    products?: ProductMasterCrafted[];
+    tools?: string[];
+    sound?: string;
+    ownership?: Record<string, unknown>;
+    ingredientsInspection?: 0 | 1;
+    productInspection?: 0 | 1;
+    img?: string;
+}
+
+/**
+ * @typedef {Object} ComponentMasterCrafted
+ * @property {string} [id=undefined]
+ * @property {string} [uuid=undefined]
+ * @property {number} [quantity=undefined]
+ * @property {string} [name=undefined]
+ */
+interface ComponentMasterCrafted {
+    id?: string;
+    uuid?: string;
+    quantity?: number;
+    name?: string;
+}
+
+/**
+ * @typedef {Object} IngredientMasterCrafted
+ * @property {string} id
+ * @property {string} [name=null]
+ * @property {ComponentMasterCrafted[]} [components=[]]
+ * @property {RecipeMasterCrafted} [recipe=null]
+ */
+interface IngredientMasterCrafted {
+    id: string;
+    name?: string;
+    components?: ComponentMasterCrafted[];
+    recipe?: RecipeMasterCrafted;
+}
+
+/**
+ * @typedef {Object} ProductMasterCrafted
+ * @property {string} id
+ * @property {string} name
+ * @property {ComponentMasterCrafted[]} components
+ */
+interface ProductMasterCrafted {
+    id: string;
+    name: string;
+    components: ComponentMasterCrafted[];
+}
+
+
+
+class MasterCraftedAPI extends DefaultRecipeAPI implements RecipeAPI {
+
+    constructor({
+        notificationService,
+        localizationService,
+        recipeValidator,
+        recipeStore,
+        identityFactory,
+        gameProvider,
+        documentManager,
+    }: {
+        notificationService: NotificationService;
+        localizationService: LocalizationService;
+        recipeValidator: RecipeValidator;
+        recipeStore: EntityDataStore<RecipeJson, Recipe>;
+        identityFactory: IdentityFactory;
+        gameProvider: GameProvider;
+        documentManager: DocumentManager;
+    }) {
+        super({
+            notificationService,
+            localizationService,
+            recipeValidator,
+            recipeStore,
+            identityFactory,
+            gameProvider,
+            documentManager
+        })
+    }
+
+    async importRecipeBookFile(recipeBookMasterCraftedPathToJson:string): Promise<Recipe[]> {
+        let recipeBookMasterCrafted: RecipeBookMasterCrafted = JSON.parse(fs.readFileSync(recipeBookMasterCraftedPathToJson, 'utf-8'));
+        return await this.importRecipeBook(recipeBookMasterCrafted);
+    }
+
+    async importRecipeBookJson(recipeBookMasterCraftedJson:string): Promise<Recipe[]> {
+        let recipeBookMasterCrafted: RecipeBookMasterCrafted = JSON.parse(recipeBookMasterCraftedJson);
+        return await this.importRecipeBook(recipeBookMasterCrafted);
+    }
+
+    async importRecipeBook(recipeBookMasterCrafted: RecipeBookMasterCrafted): Promise<Recipe[]> {
```

Comment:
I do think it would be much easier to implement and follow this process if we begin with a unit test that converts `RecipeBookMasterCrafted` to a `FabricateExportModel`. I'd have a lot of questions to ask there about what we want the behaviour to be, as I don't use Mastercrafted myself and have limited understanding of how it behaves. For example, should a Recipe Book be 1:1 with a Crafting System? To illustrate why I ask the question, once I add checks to the workflow, if you wanted the same check structures on different recipe books you'd have to re-implement them in each crafting system. We don't have to have it all implemented as we'd like from day 1, but I would want to avoid building any "gotchyas" into the first implementation if we can.

Reply:
Mastercrafted to several features here a small introductory video https://www.youtube.com/watch?v=k5_7pOvUPAU, but as much as a 1:1 correlation is ideal to be able to "switch" from that module without problems, for now the thing I'm interested in is reporting just the structure of the items in the recipes.

- N Ingredients (and relative quantities)
- X Results ( and relative quantities)
- (Optional) The tool you need yo have for this

Here you can find some json example to check out https://github.com/p4535992/foundryvtt-potion-crafting-and-gathering/tree/master/src/assets/recipes ,